ext/drreg/drreg.c

```diff
@@ -1831,115 +1831,198 @@ drreg_is_instr_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT
-     * used to avoid slot conflicts in multi-phase uses of drreg (i#3823). Perhaps
-     * that metadata can be used in this restore state code.
+    /* This routine needs to be robust enough to handle the many tricky corner
+     * cases that can arise in drreg use: gpr/aflags re-spills (to same or
+     * different slot) after app instr write, multiple reg restores due to app
+     * instr read, aflags spilled to reg instead of slot (for x86 only), lazy
+     * restores, multi-phase use of drreg leading to nested or overlapping
+     * spill regions, ...
+     * The strategy we follow here is to trace the app value of aflags and gpr
+     * as it is moved between spill slots and regs. At times the app value may
+     * be present in the gpr/aflags and spill slot both, but later either one
+     * of the two may be clobbered by a tool write or another spill
+     * respectively. At the end, we restore all live gprs/aflags that do not
+     * have their native app value.
      */
-    uint spilled_to[DR_NUM_GPR_REGS];
-    uint spilled_to_aflags = MAX_SPILLS;
+    uint gpr_spill_slot[DR_NUM_GPR_REGS];
+    uint aflags_spill_slot;
+    reg_id_t aflags_spill_reg;
+    bool gpr_native[DR_NUM_GPR_REGS];
+    bool aflags_native;
+    int last_opcode = 0;
+
     reg_id_t reg;
-    instr_t inst;
-    byte *prev_pc, *pc = info->fragment_info.cache_start_pc;
+    instr_t *inst;
+    byte *pc = info->fragment_info.cache_start_pc;
     uint offs;
     bool spill;
-#ifdef X86
-    bool prev_xax_spill = false;
-    bool aflags_in_xax = false;
-#endif
     uint slot;
     if (pc == NULL)
         return true; /* fault not in cache */
-    for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++)
-        spilled_to[GPR_IDX(reg)] = MAX_SPILLS;
+
+    /* At beginning of bb, all gprs and aflags have their native app value. */
+    for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++) {
+        gpr_spill_slot[GPR_IDX(reg)] = MAX_SPILLS;
+        gpr_native[GPR_IDX(reg)] = true;
+    }
+    aflags_spill_slot = MAX_SPILLS;
+    aflags_spill_reg = DR_REG_NULL;
+    aflags_native = true;
+
     LOG(drcontext, DR_LOG_ALL, 3,
         "%s: processing fault @" PFX ": decoding from " PFX "\n", __FUNCTION__,
         info->raw_mcontext->pc, pc);
-    instr_init(drcontext, &inst);
-    while (pc < info->raw_mcontext->pc) {
-        instr_reset(drcontext, &inst);
-        prev_pc = pc;
-        pc = decode(drcontext, pc, &inst);
-
-        if (is_our_spill_or_restore(drcontext, &inst, &spill, &reg, &slot, &offs)) {
-            LOG(drcontext, DR_LOG_ALL, 3,
-                "%s @" PFX " found %s to %s offs=0x%x => slot %d\n", __FUNCTION__,
-                prev_pc, spill ? "spill" : "restore", get_register_name(reg), offs, slot);
+    /* XXX: if ilist not available, fall back to some best-effort logic. */
+    ASSERT(info->fragment_info.ilist != NULL, "ilist required for state restoration");
```

Comment:
@derekbruening Do you think it's worth keeping the old restore state logic for cases when ilist isn't available? Just so that we may not break clients using the store-translations options.

Reply:
Yes, it is not just for client choices: core DR has to store translations for pending-deletion and selfmod.  So we can't just assert here.  Taking a selfmod test and adding a fault should hit this assert.

Maybe best to split out separate functions for the ilist vs decode (or possibly store data: xref #4937), unless they can be interleaved somehow, or the decode case builds an ilist w/ the same metadata and uses the same code in the end or sthg.

Reply:
Added two paths: ilist-absent invokes the old restore logic, and ilist-present invokes the new one; seemed cleaner to separate out and most of the code is different.

Also added another test in drreg-test that returns DR_EMIT_STORE_TRANSLATIONS to invoke the old code.